### DISCUSSION TITLE
Why are there extra features?

### DISCUSSION THREAD

**Discussion by tesujimath** [nushell/nushell#11319] [General]
I just discovered that there are extra features, after a bit of frustration trying to get `format` to work.  I found my Nushell didn't have the extra features.  Since I'm using Nix, I was able to side-step this by switching from the `nushell` package to `nushellFull`.

I didn't find much info on extra features.  Is this a mechanism for stabilisation?  Is there another reason why such things are not the default?

It seems to me that there's a downside of having extra features which may or may not be available, in fragmentation of the language.  These features can't be relied upon to be available, which makes them unusable for public scripts.

What am I missing?

**Comment by fdncred** [comment]
I think the main point on this is that there are over 400 commands and handful of people to support them. So, we created an extra feature that allows some more rarely used commands or commands that may not be as stable to be offloaded there. There's also a dataframe feature.

I always build with --features=extra,dataframe anyway because I also always want to have all the commands.

```nushell
❯ scope commands | where is_custom == false and is_builtin == true and is_plugin == false | length
491
```

**Comment by 101313** [comment]
I think given its size dataframe is rightfully excluded from the defaults but its likely a good idea to include extras by default

**Reply by tesujimath** [reply]
I wonder if there's a better way to do this?

I do understand the desire to not force everyone to have the full-fat version. And it seems reasonable to keep some unstable features back until they're ready (which won't happen until they get sufficient use and feedback, so they need to be available somehow).

But how will we differentiate between full and lite in scripts? Currently it's not possible for the script interpreter shebang line to ask for what it needs, because both variants are called nu. No I didn't mean that nu, I meant the other one!

Do we need two separate names to install these under? Or is there another way for scripts to claim what they need?

**Reply by tesujimath** [reply]
Maybe what I'm really looking for is support for libraries (different from plugins, because you may not want to pay the cost of the pipe). Then the core could be lean and mean.

Sorry if this has already been discussed and rejected, I know I'm fairly late to the party.

**Reply by fdncred** [reply]
This is one way you can tell if nushell was compiled with extra features.
```nushell
version | get features | 'extra' in $in
```
Here's another
```nushell
(version).features | str contains extra
```
The point is that features are listed in the `version` command.
